Modules/_ctypes/_ctypes.c

```diff
@@ -956,32 +956,51 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
         return NULL;
     }
 
+#undef USE_DLERROR
 #ifdef MS_WIN32
     Py_BEGIN_ALLOW_THREADS
     address = (void *)GetProcAddress(handle, name);
     Py_END_ALLOW_THREADS
-    if (!address) {
-        PyErr_Format(PyExc_ValueError,
-                     "symbol '%s' not found",
-                     name);
-        return NULL;
-    }
 #else
+    #ifdef __CYGWIN__
+        // dlerror() isn't very helpful on cygwin
+    #else
+        #define USE_DLERROR
+        /* dlerror() always returns the latest error.
+         *
+         * Clear the previous value before calling dlsym(),
+         * to ensure we can tell if our call resulted in an error.
+         */
+        (void)dlerror();
+    #endif
     address = (void *)dlsym(handle, name);
-    if (!address) {
-#ifdef __CYGWIN__
-/* dlerror() isn't very helpful on cygwin */
-        PyErr_Format(PyExc_ValueError,
-                     "symbol '%s' not found",
-                     name);
-#else
-        PyErr_SetString(PyExc_ValueError, dlerror());
 #endif
-        return NULL;
+
+    if (address) {
+        ctypes_state *st = get_module_state_by_def(Py_TYPE(type));
+        return PyCData_AtAddress(st, type, address);
+    }
+
+#ifdef USE_DLERROR
+    // This assumes the error message is UTF-8 (or ASCII).
+    // Investigate if this can cause problems.
+    const char *dlerr = dlerror();
+    if (dlerr) {
+        PyObject *message = PyUnicode_DecodeLocale(dlerr, "strict");
+        if (message) {
+            PyErr_SetObject(PyExc_ValueError, message);
+            return NULL;
+        }
+        // Ignore errors from converting the message to str
+        PyErr_Clear();
```

Comment:
This doesn't seem right, why? We might want to use `PyErr_WriteUnraisable` here.

Reply:
I cannot argue for or against this, as I don't have the context.

Am I expected to get the context, or should I wait for someone else (the other reviewers) to answer?

Reply:
Did @encukou tell you to do it this way? (If so, I'll yield.)

Reply:
I did, but before I saw the precedent of `surrogateescape` (in handling `errno`/`strerror` for `OSError`, in `errors.c`).
With `surrogateescape`, an exception raised here will should truly unexpected (e.g. a MemoryError); we can raise that instead of the generic ValueError.
```suggestion
        return NULL;
```

Reply:
So, what's the verdict? I'm not 100% in sync.

Reply:
Use `surrogateescape` and then propagate errors when decoding to the caller.

Reply:
Done and pushed, PTAL to confirm in case I misunderstood something

Reply:
Sorry, this PR got lost in notifications. I'll review this again tomorrow. If I forget, please ping me!